### ISSUE TITLE
feature request: upsampling

### ISSUE THREAD

**Issue by lzhengchun** [intel/mkl-dnn#563]
like [tf.keras.layers.UpSampling2D](https://www.tensorflow.org/api_docs/python/tf/keras/layers/UpSampling2D)
even just "nearest" method. this is useful for some image restoration problem like denoise, super resolution; and segmentation

**Comment by lzhengchun** [comment]
when upsampling primitive is not available, i found these lines of code to implement one using nearest. any thoughts on doing similar thing using DNNL?

`
def UpSamplingCustom2D(scale=(2, 2)):
    if isinstance(scale, int):
        scale = (scale, scale)
    def upsampling(x):
        shape = x.shape
        x = tf.keras.layers.Concatenate(-2)([x] * scale[0])
        x = tf.keras.layers.Reshape([shape[1] * scale[0], shape[2], shape[3]])(x)
        x = tf.keras.layers.Concatenate(-1)([x] * scale[1])
        x = tf.keras.layers.Reshape([shape[1] * scale[0], shape[2] * scale[1], shape[3]])(x)
        return x
    return upsampling
`

**Comment by vpirogov** [comment]
Hi @lzhengchun,

Currently DNNL does not provide support for upsampling. In cases like that applications are expected to implement the functionality in standard programming model for the device (C++ for CPU, OpenCL for GPU). In case when scaling factors are integer maximum pooling backpropagation will do exactly what need.

The upsampling primitive is a reasonable candidate to consider for the library though. Tricky part is defining operation semantics in the way that would be generally useful.

**Comment by lzhengchun** [comment]
Thanks @vpirogov My scaling factor is 2 on both H and W dimension. Could you give a little bit instruction on how to achieve using maximum pooling backpropagation?

**Comment by vpirogov** [comment]
Maximum pooling gradient with 2x2 kernel and 2x2 stride will produce the required output. [CNN training example](https://github.com/intel/mkl-dnn/blob/master/examples/cnn_training_f32.cpp#L250) demonstrates max pooling backpropagation. 

I was not able to create backward pooling without workspace though. Summoning @emfomenk for help.

**Comment by lzhengchun** [comment]
@vpirogov I saw that backward must have associate forward primitive description. I only need forward for my inference. any thought @emfomenk 

Also, I plan to use **deconvolution**, and want **same** padding as it in tensorflow to make a perfect 2x scaling up. it works when I set kernel size as 2, stride as 2 and padding as {0, 0}.

But when I try to use kernel size of 4, I do not succeed on setting  the padding, I tried {-1,-1}, it does not work. if padding is {0,0}, kernel size is 4 and stride at 2, the output will be 2x + 2 (x is the original size). does DNNL support negative padding? I could see any documentation explains this and there is no example as well. 

I only found a test (under gtest), but do not understand what does _right_padding_ mean.

**Comment by emfomenk** [comment]
> Maximum pooling gradient with 2x2 kernel and 2x2 stride will produce the required output.

Not sure about this. Backward max-pooling produces lots of zero -- basically at places that are not maximums (this is because these elements do not contribute to the error). Hence:
```
// backward max pooling windows=2x2, stride=2x2
                [ 1 0 2 0 ]
 [ 1 2 ]        [ 0 0 0 0 ]
 [ 3 4 ]   -->  [ 3 0 4 0 ]
                [ 0 0 0 0 ]
```

What might work is backward average pooling: since forward takes the whole window and average it to the output, on backward it would basically spread the gradient. The only problem here, is that it would also divide by the window size:

```
// backward max pooling windows=2x2, stride=2x2
                [ .25 .25 .5 .5 ]
 [ 1 2 ]        [ .25 .25 .5 .5 ]
 [ 3 4 ]   -->  [ .75 .75 1. 1. ]
                [ .75 .75 1. 1. ]
```

So, to get the _right_ answer one would need to re-scale the result by 4.
Anyways, this looks way too hacky. It would probably be better to provide a separate primitive for that.

> But when I try to use kernel size of 4, I do not succeed on setting the padding, I tried {-1,-1}, it does not work. if padding is {0,0}, kernel size is 4 and stride at 2, the output will be 2x + 2 (x is the original size). does DNNL support negative padding? or how can I achieve perfect 2x upscale using deconv?

Padding cannot be negative. The formula for convolution shape is:
```
oh = (ih + left_padding + right_padding - kh) / stride + 1
// oh -- output height
// ih -- input height
```

For deconvolution oh is an input, and ih is an output. But that doesn't matter much if you want to have the same size of input and output. For odd kernel size the both left and right padding should be the same and be equal `kernel / 2` (round down). For even kernel size the paddings should be:
- Either `(kernel / 2, kernel / 2 + 1)`,
- Or `(kernel / 2 + 1, kernel / 2)`

Picture:
```
src:   [a b c d]  // 4 points, but might be odd as well, does't matter
kernel [+ +]      // 2 points, even number

[a b c d]  
 + +        | output[0]
   + +      | output[1]
     + +    | output[2]
       + +  | output[3] -- need padding to produce this point
         ^
         |- padding
```

> I could not find the documentation explains this and there is no example as well. I only found a test (under gtest), but do not understand what does right_padding mean.

Right and left paddings are conventions. Left padding means padding from the front / beginning / left side of the area, while the right padding means the back / end / right padding:
- 1D case (W):
    - left_padding = {left_padding}
    - right_padding = {right_padding}
- 2D case (HW):
    - left_padding = {top_padding, left_padding},
    - right_padding = {bottom_padding, right_padding}
- 3D case (DHW):
    - left_padding = {front_padding, top_padding, left_padding},
    - right_padding = {back_padding, bottom_padding, right_padding}

**Comment by lzhengchun** [comment]
Thansk @emfomenk. At the end i used deconv with kernel size of 2 and stride of 2 as well, it gives me perfect 2x upscaling. 

Very unfortunate, after a day of debug, I still cannot get the same results as tensorflow (I used tf to train my model). Interestingly, in order to test, i set weights to 1 and bias to 0, it gives the same results as tf.

But when I use real weights, it doesnot give me correct results any more.  I opened another [issue](https://github.com/intel/mkl-dnn/issues/568) to seek for help 

Could you please review my code? 

Thanks

**Comment by atafra** [comment]
A nearest-neighbor upsampling primitive would be very useful for Open Image Denoise as well.

**Comment by rsdubtso** [comment]
@irinasok, can you please post which options for upsampling API we are considering now?

**Comment by irinasok** [comment]
@rsdubtso, sure!
@atafra, @lzhengchun we are working on the new primitive to support **resize** operation that will do upsampling if user passes factor > 1 and downsampling if factor < 1
The API will look something like this:
```
    memory::dims src_tz = {batch, channels, 10, 10};
    auto src_md = memory::desc({src_tz}, dt::f32, tag::nchw);
    memory::dims dst_tz = {batch, channels, 20, 20};
    auto dst_md = memory::desc({dst_tz}, dt::f32, tag::nchw);
    std::vector<float> factors = {2, 2};

    // Create resize primitive
    auto resize_desc = resize_forward::desc(prop_kind::forward_inference,
            algorithm::resize_nearest, factors, src_md, dst_md);
    auto resize_pd = resize_forward::primitive_desc(resize_desc, eng);
```

New primitive will support:

- 3D, 4D and 5D input and output data tensors in plain (ncw, nchw, ncdhw, nhwc, etc) and blocked (nChw16c, nChw8c) memory layouts. 

- 2 interpolation methods: Nearest Neighbor and Linear (bilinear for 4D tensor, trilinear for 5D tensor). 

- Forward and backward by data propagation.

- Different scaling factors in each spacial dimension, of float type.

There are some restrictions:

- input and output tensor must have the same data type and layout (if input is f32 with nchw layout, output will also be f32 with nchw)

- we are likely to start with support of f32 and bf16 data types only and will extend with int8 by request. 

I also want to point out that there are [many different approaches](https://jricheimer.github.io/tensorflow/2019/02/11/resize-confusion/) to interpolation [that caused confusion for framework users](https://hackernoon.com/how-tensorflows-tf-image-resize-stole-60-days-of-my-life-aba5eb093f35). In DNNL we elected to follow the OpenCV and TF 2.0 approach, that aligns input and output pictures in the center for all interpolation methods.  

Please let me know if any of the above doesn't look right to you, while we still have time to change the API.

**Comment by lzhengchun** [comment]
@irinasok that's exactly what we hope to have. Is it publicly available now?

**Comment by vpirogov** [comment]
@lzhengchun, we are still working on the implementation. I'll let you know once it lands to master.

**Comment by xu78guo6** [comment]
@vpirogov , may i kindly ask when to release v1.2, really need resize bilinear primitive to implement a CNN.

**Comment by emfomenk** [comment]
Hi @xu78guo6,

According to the [milestone page](https://github.com/intel/mkl-dnn/milestone/1) the release is expected in Jan'20. But the implementation should definitely appear earlier than that.

**Comment by nafest** [comment]
@irinasok , the ONNX specification has a parameter 'coordinate_transformation_mode' which seems to be necessary to have a resize operation that is compatible with a broad range of available deep learning frameworks: https://github.com/onnx/onnx/blob/master/docs/Operators.md#Resize

I have my own optimized implementation for Resize (only supporting the most common layouts nChw8c and nChw16c) and supporting different modes really isn't that hard. Please consider to add this as parameter to DNNL as well.

**Comment by irinasok** [comment]
Hi @nafest, the direction we are taking with resize operation is to support only coordinate_transformation_mode="half_pixel". This way we will support TF 2.0 resize and Pytorch interpolate for most cases. The main reason is that frameworks tend to switch to this approach as the one that makes most sense from user perspective (at least it seems like it from this [thread](https://github.com/tensorflow/tensorflow/issues/6720)).

**Comment by vpirogov** [comment]
Implemented in 16b9a5d3abfd5e4b900f07fbd1ba09662952f38f.
